server/README.md

```diff
@@ -0,0 +1,236 @@
+any client accessing the block after T2 will reject it given it appears to have
+been created after Alice lost her access to the realm !
+
+## Causality & concurrency model
+
+From the server point of view, there is basically four different kinds of data:
+
+- Certificates
+- Blocks
+- Vlobs
+- Invitations
+
+### Certificates
+
+Any change done by a client that should be visible by the server is provided as a
+certificate (i.e. a certificate is an unencrypted document signed by the client's device).
+
+A very simple way to ensure causality and concurrency on certificates would be:
+
+- Have a single big lock in the server that must be taken before any attempt to
+  add a certificate (handles concurrency).
+- Ensure the new certificate's timestamp is newer than any existing certificate
+  (handles causality).
+
+However this is pretty wasteful given most certificates are unrelated (e.g. certif A
+add a new user, certif B does a key rotation in a realm). So instead we reduce the
+size of the lock by introducing the concept of topic.
+
+Each type of certificate is part of one of the following topics:
+
+- `common`
+- `sequester`, which depends on `common`
+- `realm`, which depends on `common` and uses the realm ID as discriminant
+- `shamir_recovery`, which depends on `common` and use the ID of the user to recover as discriminant
+
+So the actual rule used in the server is:
+
+- Have lock in the server for each couple topic + discriminant
+- Take the lock for the topic + discriminant couple the certificate we want to add
+  belongs to AND to other topics our topic may depend on (handles concurrency).
+- Ensure the new certificate's timestamp is newer than any existing certificate
+  in the topics we have locked (handles causality).
+
+> Note: Also organizations are totally isolated from each other on the server, so
+> in a way the organization ID is an aditional discriminant on each certificate topic.
+
+### Blocks
+
+Blocks are encrypted piece of data that belong to a given realm and cannot
+be modified.
+
+This makes concurrency & causality pretty simple:
+
+- Each block is considered independent, hence there is no need for lock dedicated to blocks.
+- A block is identified by a client-provided UUID, then there is no risk for double-creation
+  under concurrency (a trivial unicity check on insertion prevents this).
+- Block creation date must be more recent than the last certificate in `common` topic and
+  in the `realm` topic corresponding to the block's realm.
+
+> Notes:
+>
+> - It could be tempting to only lock the `realm` topic (given it's this topic that contains
+>   the certificate describing the role of a user in the realm). However a revoked user is also
+>   forbidden from adding data, hence making it mandatory to also lock the `common` topic here.
+> - Uploading blocks (and vlobs) are the most common task on the server, so we may want to
+>   uptomise it in the future to avoid having to lock two topics.
```

Comment:
That sounds quite bad. The proposed strategy should already address this problem IMO.

Reply:
I've run dictionary check and corrected plenty !

Reply:
I didn't mean the syntax, I mean that maybe it's not something to optimize in the future but deal with now in this design document.